### ISSUE TITLE
Implement subset of File System Access API

### ISSUE THREAD

**Issue by Zipdox2** [mozilla/standards-positions#738]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: File System Access API
* Specification or proposal URL:  https://fs.spec.whatwg.org/ and https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API
* Caniuse.com URL (optional): https://caniuse.com/native-filesystem-api
* Bugzilla URL (optional): https://bugzilla.mozilla.org/show_bug.cgi?id=1246236
* Mozillians who can provide input (optional): 

### Other information

This was previously turned down (https://github.com/mozilla/standards-positions/issues/154 and https://github.com/Daasin/standards-positions/commit/91a1e8c2e8b3b16a9a166e86bf7fefb787f56502), however, I took notice of the "Mozilla could be supportive of parts" part. I think that this API could provide some specific very important functionality that is not harmful to the user. Please don't close this issue without reading it thoroughly and considering my case.

I think the primary concern with the API is the `window.showDirectoryPicker()` method and the associated `FileSystemDirectoryHandle`. Selecting a directory would allow a website to create and delete any number of files within the directory, and I agree that this is very dangerous, and should not be implemented (or disabled by default).

But I believe that `window.showOpenFilePicker()` and `window.showSaveFilePicker()` and the associated `FileSystemFileHandle` don't create any increased risk to the user than already exists today. This allows only access to one file. Opening and saving files is already possible without this API, but it doesn't provide the same functionality. The only possible problem is a website writing to a file that the user opened, because opening a file traditionally only implies reading, but this can be solved by simply prompting the user. Of course a prompt wouldn't be necessary for files created with showSaveFilePicker, since that already implies writing.

Why I think this is important is because FileSystemFileHandle allows streamed downloads for client-generated files. Currently, doing this requires a hack with service workers, which are not accessible in private mode and require a worker script to be hosted (see https://github.com/jimmywarting/StreamSaver.js). Without this you're limited to downloading files that fit within memory. Additionally, it allows seeking within a file and truncating the file.

**TL;DR:** Implement only `FileSystemFileHandle` and not `FileSystemDirectoryHandle`.

**Comment by jimmywarting** [comment]
> Opening and saving files is already possible without this API

Same could be said for reading folders as well. (but not writing)
You can drag and drop a folder into a website and call `item.webkitGetAsEntry()`

Clients can update the content in that folder and website can read the newly updated files and folder in that directory with the old entry api (as long as you have not GC it)
(And you can't store this entry in IndexedDB either)

You can also get all files (recursively) by using `<input type="file" webkitdirectory>` (which also is available in FF)

So in some since i think that parts of `showDirectoryPicker()` could be allowed as well.
...at least for just reading the files for the duration of that browser session. (so you would not be able to create new files/folders so that would instead reject the promises)

**Comment by Zipdox2** [comment]
It's the writing part that's actually important here.

**Comment by Zipdox2** [comment]
And the download object:

{{APIRef("HTML DOM")}}

The **`Download`** object represents an ongoing or finished download.

Downloads can tell you about how far the file has been downloaded, how big it is (if not a stream), its file name, and when it finishes or aborts.

## Instance properties

- {{DOMxRef("Download.prototype.size")}} {{ReadOnlyInline}}
  - : The size, in bytes, of the data being downloaded. If it is a stream, this will be `undefined`.
- {{DOMxRef("Download.prototype.progress")}} {{ReadOnlyInline}}
  - : The size, in bytes, of the data that has been downloaded so far.
- {{DOMxRef("Download.prototype.name")}} {{ReadOnlyInline}}
  - : A string indicating the name of the file being downloaded.

## Instance methods

- {{DOMxRef("Download.prototype.abort()")}}
  - : Aborts the download and deletes the downloaded file. If the download was created from a stream, it will be unlocked.
- {{DOMxRef("Download.prototype.finish()")}}
  - : Stops downloading the file and closes it up. If the download was created from a stream, it will be unlocked.

## Events

- {{DOMxRef("Download.progress_event", "progress")}}
  - : Fired when the downloaded progresses.
- {{DOMxRef("Download.finish_event", "finish")}}
  - : Fired when the download has finished. If the download was created from a stream, the stream has closed.
- {{DOMxRef("Download.error_event", "error")}}
  - : Fired when there is an error with downloading the file. This includes the user canceling the download.

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

**Comment by martinthomson** [comment]
Thanks for raising the issue, but after discussion, we've concluded that we need to defer this question.

The idea that there might be elements of the specification that are safe to use is not one that we've explored fully.  That said, we try to avoid taking specifications à la carte, because that makes it harder to manage interoperability, especially when it results in more complicated feature detection.

If we were to consider this idea, we'd like to see a more thorough analysis.  That means a crisp definition of what is included, the implications of those choices, and more extensive work on understanding all of the implications (privacy, security, usability, alternatives, feature detection, interaction with other features, use cases and demand).  There are potentially good mitigations already in place for some of the risks that have been identified (like a web site modifying a file that is actively in use by another program), but we'd need to see a more comprehensive treatment.  Also, it's not clear that this would be responsive to the requests from web developers.

I understand that this is asking a fair bit, but it is generally accepted that the proponents of new web features do this sort of work.  Opening an issue on this repository isn't the place I'd recommend taking that work, either.  The WHATWG processes for new proposals are better suited to handling this sort of thing.

**Comment by goldwaving** [comment]
Implementing some form of a File System Access API is inevitable.  There is no other way to efficiently modify large files or sets of files (audio, video, AI models, etc.).

**Comment by Zipdox2** [comment]
> Implementing some form of a File System Access API is inevitable. There is no other way to efficiently modify large files or sets of files (audio, video, AI models, etc.).

I agree.

**Comment by Zipdox2** [comment]
Can we bring this up again?

**Comment by Zipdox2** [comment]
I'm writing an add-on that needs to stream a download of a file that's generated in real-time. This may be possible using StreamSaver.js but that requires a hack with a service worker, which I'm not sure is possible in an add-on. Having showOpenFilePicker would make my life infinitely easier.

**Comment by Zipdox2** [comment]
@martinthomson

**Comment by kingIZZZY** [comment]
+1 need `window.showSaveFilePicker()` for all browsers 🙏

**Comment by jesup** [comment]
To reiterate what Martin said in Jan 2023, the right way to move forward on this is to push for a concrete specification that's safe for users in the WHATWG process.  This has been done before - OPFS is (largely) a subset of the File System API that's safe and allows high-performance reading and writing of local files, though in a sandbox, not the general filesystem.

**Comment by Zipdox2** [comment]
OPFS is very useful, and I do use it, but it would require using twice as much storage space to save a single file. Let's suppose I'm developing an encrypted file sharing site with client side decryption, and I use OPFS. A user downloads a 100GB file. This file first gets written to OPFS. Once that's complete, I use getFile() and URL.createObjectURL() and download the file. This eats up another 100GB of storage space. Now it's using 200GB. Only once the download is complete can I remove the OPFS file. See the problem?

As for a new specification in the WHATWG process, there have been multiple proposals (https://github.com/whatwg/html/issues/4148 and https://github.com/WICG/proposals/issues/111) for a download API and none of them have gained any traction. This is exacerbated by the fact that they are redundant since navigator.showSaveFilePicker() exists.
